draft-ietf-jsonpath-base.md

```diff
@@ -1006,55 +1005,66 @@ The current item is selected if and only if the result is `true`.
-               "<"  / ">"  /                          ; operators
-               "<=" / ">="
-number       = int [ frac ] [ exp ]                   ; decimal number
-frac         = "." 1*DIGIT                            ; decimal fraction
-exp          = "e" [ "-" / "+" ] 1*DIGIT              ; decimal exponent
-true         = %x74.72.75.65                          ; true
-false        = %x66.61.6c.73.65                       ; false
-null         = %x6e.75.6c.6c                          ; null
+comparable   = json-value / path                    ; JSON value or path value
+json-value   = number / string-literal /            ; primitive
+               true / false / null /                ; value or
+               object / array /                     ; structured value or
+               path                                 ; path value
+comp-op      = "==" / "!=" /                        ; equality or
+               "<"  / ">"  /                        ; relational
+               "<=" / ">="                          ; operator
+number       = int [ frac ] [ exp ]                 ; decimal number
+frac         = "." 1*DIGIT                          ; decimal fraction
+exp          = "e" [ "-" / "+" ] 1*DIGIT            ; decimal exponent
+true         = %x74.72.75.65                        ; true
+false        = %x66.61.6c.73.65                     ; false
+null         = %x6e.75.6c.6c                        ; null
+
+object       = "{" S [ members ] S "}"              ; { ... }
+members      = member *(S "," S member) 
+member       = json-string S ":" S json-value       ; "name": value
+    
+array        = "[" S [ elements ] S "]"             ; [ ... ]
+elements     =  json-value *(S "," S json-value )
 
 regex-expr   = (path / string-literal) S regex-op S regex
-regex-op     = "=~"                                   ; regular expression match
+regex-op     = "=~"                                 ; regular expression match
 regex        = <TO BE DEFINED>
 ~~~~
 
 Notes:
 
 * Parentheses can be used with `boolean-expr` for grouping. So filter selection syntax in the original proposal `[?(<expr>)]` is naturally contained in the current lean syntax `[?<expr>]` as a special case.
-* Comparisons are restricted to primitive values (such as number, string, `true`, `false`, `null`). Comparisons with complex values will fail, i.e. no selection occurs.
-<!-- issue: comparison with structured value -->
 * Types are not implicitly converted in comparisons.
   So `"13 == '13'"` selects no node.
-* A member or element value by itself in a Boolean context is
-  interpreted as `false` only if it does not exist.
+* Relational operators (`>`, `<`, `>=`, `<=`) are working on primitive number values exclusively. 
+Comparison with a non-number value(s) fails, i.e. no selection occurs.
+* Equality operators (`==`, `!=`) on primitive and/or structured values are working as expected, i.e. selection occurs in case of deep equality/inequality.
```

Comment:
What happens when a path is compared to an array and the path produces a nodelist which happens to be the same as the contents of the array, in the same order?

Reply:
I have in mind to allow only a path referencing a single array having the same content as the array to be compared. Nodelists are to be discussed ...

Reply:
Perhaps Singular Paths (https://github.com/ietf-wg-jsonpath/draft-ietf-jsonpath-base/pull/164) helps with this.